src/neuron/container/soa_container.hpp

```diff
@@ -742,40 +766,63 @@ struct soa {
         (std::get<tag_index_v<Tags>>(m_data).for_all_vectors(callable), ...);
     }
 
+    /**
+     * @brief Record that a state_token was copied.
+     *
+     * This should only be called from the copy constructor of a state_token,
+     * so m_frozen_count should already be non-zero.
+     */
+    void increase_frozen_count() {
+        std::lock_guard _{m_mut};
+        assert(m_frozen_count);
+        ++m_frozen_count;
+    }
+
     /**
      * @brief Flag that the storage is no longer frozen.
      *
      * This is called from the destructor of state_token.
      */
     void decrease_frozen_count() {
+        std::lock_guard _{m_mut};
         assert(m_frozen_count);
         --m_frozen_count;
     }
 
   public:
     /**
-     * @brief Return type of get_sorted_token()
+     * @brief Return type of issue_frozen_token()
      */
-    using sorted_token_type = state_token<Storage>;
+    using frozen_token_type = state_token<Storage>;
 
     /**
-     * @brief Mark the container as sorted and return a token guaranteeing that.
+     * @brief Create a token guaranteeing the container is in "frozen" state.
+     *
+     * This does *not* modify the "sorted" flag on the container.
+     *
+     * The token type is copiable but not default constructible. There is no
+     * need to check if a given instance of the token type is "valid"; if a
+     * token is held then the container is guaranteed to be frozen.
+     *
+     * The tokens returned by this function are reference counted; the
+     * container will be frozen for as long as any token is alive.
+     *
+     * Methods such as apply_reverse_permutation() take a non-const reference
+     * to one of these tokens. This is because a non-const token to a container
+     * with a token reference count of exactly one has an elevated status: the
+     * holder can lend it out to methods such as apply_reverse_permutation() to
+     * authorize specific pointer-invaliding operations. This is useful for
```

Comment:
Yes, in the sense that normally if one holds a token then, so long as that token is alive, you can assume that data will not move around in memory:
```c++
void foo(frozen_token_type const&) {
  auto* ptr = &handle_to_row.a();
  /* anything */
  assert(ptr == &handle_to_row.a();
}
```
because we know there is a const token alive during `/* anything */`, we know that the assertion will always hold (because any operations that attempt to break that condition, like adding a new row, will fail).

In the similar-but-different case:
```
void foo(frozen_token_type& token) {
  auto* ptr = &handle_to_row.a();
  /* care needed here */
  assert(ptr == &handle_to_row.a(); // !!!
}
```
we still know that the assertion will not fail due to unrelated code/threads, but because there is a non-const token alive and available here, we need to pay attention to what happens in `/* care needed here */`. Specifically, if that code includes:
```c++
  soa_data.apply_reverse_permutation(perm, token);
```
then the assertion can easily be made to fail.

Note that, because we do not have a fully-featured system of read/write tokens, the call to `apply_reverse_permutation` could itself fail, if some other part of code simultaneously holds a token.